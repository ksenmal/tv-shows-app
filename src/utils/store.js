import { reactive } from "vue";
import api from "./api.js";

function compareByWeight(a, b) {
  return b.weight - a.weight;
}

export const store = reactive({
  list: [],
  isLoading: false,
  error: null,
  getList() {
    this.error = null;
    this.isLoading = true;
    return api
      .getAllShows()
      .then((resp) => {
        this.list = resp.sort(compareByWeight);
        this.error = null;
        this.isLoading = false;
      })
      .catch((error) => {
        this.error = error;
        this.isLoading = false;
        console.error(error);
      });
  },
  searchValue: "",
  setSearchValue(newValue) {
    this.searchValue = newValue;
  },
});
