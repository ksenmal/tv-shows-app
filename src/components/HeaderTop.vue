<script setup>
import { RouterLink, useRouter } from "vue-router";
import { store } from "../utils/store.js";

const router = useRouter();

function handleSearch() {
  router.push({
    path: "/search",
    query: {
      query: store.searchValue,
    },
  });
}
</script>

<template>
  <header>
    <div class="wrapper">
      <RouterLink to="/"><h1 class="green">TV Shows</h1></RouterLink>
      <div class="search">
        <input
          :value="store.searchValue"
          @input="(event) => store.setSearchValue(event.currentTarget.value)"
          @keyup.enter="handleSearch"
          placeholder="Search"
        />
        <button @click="handleSearch" :disabled="!store.searchValue">Go</button>
      </div>
    </div>
  </header>
</template>

<style scoped>
header {
  line-height: 1.5;
}
.wrapper {
  width: 100%;
  text-align: center;
}
.wrapper .search {
  margin: 1rem 0;
}
.wrapper input {
  padding: 5px 5px 5px 25px;
  border-radius: 5px;
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' class='h-6 w-6' fill='none' viewBox='0 0 24 24' stroke='%23999' stroke-width='2'%3E%3Cpath stroke-linecap='round' stroke-linejoin='round' d='M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z' /%3E%3C/svg%3E");
  background-repeat: no-repeat;
  background-position: 3px 4px;
  background-size: 1rem;
}
.wrapper button {
  padding: 4px;
  cursor: pointer;
  margin-left: 5px;
  border-radius: 4px;
}
.wrapper button:disabled {
  background-color: #b1b1b1;
  cursor: auto;
}
@media (min-width: 576px) {
  .wrapper {
    width: 100%;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    text-align: center;
  }
}
</style>
