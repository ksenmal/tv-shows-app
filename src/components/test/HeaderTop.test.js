import { shallowMount } from "@vue/test-utils";
import HeaderTop from "../HeaderTop.vue";
import { afterAll, beforeEach, describe, expect, test, vi } from "vitest";
import { store } from "../../utils/store.js";

vi.mock("vue-router", () => {
  const mockPush = vi.fn();
  return {
    RouterLink: vi.fn(),
    useRouter: vi.fn().mockImplementation(() => {
      return {
        push: mockPush,
      };
    }),
  };
});

vi.mock("../../utils/store.js", () => {
  return {
    store: {
      setSearchValue: vi.fn(),
      searchValue: "test",
    },
  };
});

describe("HeaderTop component test suite", async () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(HeaderTop);
  });

  test("Snapshot match", async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Triggering input sets search value", async () => {
    const testInput = "test Input";
    const input = wrapper.find("input");
    await input.setValue(testInput);
    await input.trigger("input");

    expect(store.setSearchValue).toHaveBeenCalledWith(testInput);
  });

  test("Trigger search input redirects to search page", async () => {
    const { useRouter } = await import("vue-router");
    const router = useRouter();
    const searchButton = wrapper.find("button");
    const input = wrapper.find("input");

    await input.setValue("test");
    await searchButton.trigger("click");

    expect(router.push).toHaveBeenCalled();
  });

  afterAll(() => {
    vi.clearAllMocks();
  });
});
