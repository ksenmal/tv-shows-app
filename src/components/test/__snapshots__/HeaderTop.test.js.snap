// Vitest Snapshot v1

exports[`HeaderTop component test suite > Snapshot match 1`] = `
"<header data-v-bfc77041=\\"\\">
  <div class=\\"wrapper\\" data-v-bfc77041=\\"\\">
    <spy-stub to=\\"/\\" data-v-bfc77041=\\"\\"></spy-stub>
    <div class=\\"search\\" data-v-bfc77041=\\"\\"><input placeholder=\\"Search\\" data-v-bfc77041=\\"\\"><button data-v-bfc77041=\\"\\">Go</button></div>
  </div>
</header>"
`;
